spigot/src/main/java/org/geysermc/floodgate/addon/data/SpigotDataHandler.java

```diff
@@ -58,8 +59,28 @@ protected void setNewIp(Channel channel, InetSocketAddress newIp) {
 
     @Override
     protected Object setHostname(Object handshakePacket, String hostname) {
-        setValue(handshakePacket, ClassNames.HANDSHAKE_HOST, hostname);
-        return handshakePacket;
+        if (ClassNames.IS_PRE_1_20_2) {
+            // 1.20.1 and below
+            setValue(handshakePacket, ClassNames.HANDSHAKE_HOST, hostname);
+
+            return handshakePacket;
+        } else {
+            // 1.20.2 and above
+            try {
+                Object[] components = new Object[]{
+                    ClassNames.HANDSHAKE_PORT.get(handshakePacket),
+                    hostname,
+                    ClassNames.HANDSHAKE_PROTOCOL.get(handshakePacket),
+                    ClassNames.HANDSHAKE_INTENTION.get(handshakePacket)
+                };
+
+                return ClassNames.HANDSHAKE_PACKET_CONSTRUCTOR.newInstance(components);
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+                e.printStackTrace();
```

Comment:
I'd just rethrow the exception if something goes wrong

Reply:
https://github.com/GeyserMC/Floodgate/pull/449/commits/55c03e3804ac57b025a9cd7bfa5f6cf58746cef3#diff-a83ed678d5742d28bade114ac629507328659171b3c3f3440f22c18066b7bdbfR79